Update advanced user/developer guides (#4905)
GenericMappingTools/gmt: MAINTENANCE.md

Information about failing tests is produced in `test/fail_count.txt` inside the build directory. For plot-based tests,
the subdirectories `test/` and `doc/scripts/` inside the build directory contain folders for each failing test. For
plot-based tests, the directory associated with each failing tests contains a `gmtest.sh` script, a `gmt.conf` file, an
alias to the test script, a PostScript file and PDF document generated by the test script, and a PNG image that shows
differences between the reference plot and new plot in magenta. In addition to these files, running the failing tests
with verbose output can be helpful for evaluating failures:

        ctest --rerun-failed --verbose

### Updating reference plots for tests

Pull requests should avoid needing to change PostScript files in the [test/](/test) and [doc/scripts/](/doc/scripts)
directories. However, if this is unavoidable, new PostScript reference files can be generated by running
`ctest -R <test-script-name>` in the build directory after following the [building guide](BUILDING.md) and the
[Configuring CMake for testing GMT](MAINTENANCE.md#configuring-cmake-for-testing-gmt) instructions. The new PostScript

> file can be copied from the appropriate subdirectory within `build/test/` or `build/doc/scipts/`.

Reviewer: ```suggestion
file can then be copied from the appropriate subdirectory within `build/test/` or `build/doc/scripts/`
to [test/](/test) or [doc/scripts/](/doc/scripts) respectively.
```

Assuming you did not delete the build directory, this is just as simple as

```
cd <path-to-gmt>
git pull
cd build
cmake --build .
cmake --build . --target install
```

CMake will detect any changes to the source files and will automatically reconfigure. If you deleted all files inside
the build directory you have to run CMake again manually.

> Note: Refer to the file `admin/bashrc_for_gmt` for useful aliases for updating the development source code

## Debugging GMT

Guides for debugging GMT are provided in the
[developer resources section](https://docs.generic-mapping-tools.org/dev/debug.html) of the GMT documentation.

## Using build and test aliases

The file [admin/bashrc_for_gmt](/admin/bashrc_for_gmt) contains useful aliases for building and testing GMT that some

> developers chose to use. New pull requests with other aliases that you find helpful are welcome. This file is version controlled, so you should copy the file to a different location in order to edit and use it. For example, use these
> commands to copy it to your home directory:

Reviewer: Small nitpick to reduce line length.

```suggestion
developers chose to use. New pull requests with other aliases that you find helpful are welcome.
This file is version controlled, so you should copy the file to a different location in order to edit and use it.
For example, use these commands to copy it to your home directory:
```

```
cd <path-to-gmt>
cp admin/bashrc_for_gmt ~/bashrc_for_gmt
```

> Here are the steps for setting up `bashrc_for_gmt` after copying it to a new location:

Reviewer: ```suggestion
Here are the steps for setting up [`bashrc_for_gmt`](/admin/bashrc_for_gmt) after copying it to a new location:
```

- If you do not have [ninja](https://ninja-build.org/) installed, you will need to change `builder=ninja` to

>   `builder=make` and `Bname="Ninja"` to `Bname="Unix Makefiles"`. It is recommended that you use ninja.

Reviewer: ```suggestion
  `builder=make` and `Bname="Ninja"` to `Bname="Unix Makefiles"`. Ninja is recommended for speeding up build times.
```

- You may need to update `pngview=open` and `pngview=open` depending on your preferred program for viewing files.
- Optionally, change `ncores=4` to the number of cores to use for building and running tests.
- Change `MATLAB=/Applications/MATLAB_R2019a.app` to the path for your version of the MATLAB app.

> - Set `REPO_DIR` to the path that contains the GMT repository.

Reviewer: ```suggestion
- Set `REPO_DIR` to the path that contains the local `git clone` copy of the GMT repository.
```

- Set `DATA_DIR` to the path that contains the folders `dcw-gmt-1.1.4/` and `gshhg-gmt-2.3.7/` for the dcw and gshhg
  datasets respectively. If these folders are not located in the same path, you can instead delete L25
  (`DATA_DIR=<path to directory containing GSHHG and DCW>`) and set the individual paths to the GSHHG and DCW source
  in L30 (`export GMT_GSHHG_SOURCE=${DATA_DIR}/gshhg-gmt-2.3.7`) and

>   L31 (`export GMT_DCW_SOURCE=${DATA_DIR}/dcw-gmt-1.1.4`).

Reviewer: I worry that the line numbers may change in the future, so maybe remove the L25/L30/L31? What do you think, keep or remove?

Author: Yes, I removed the line numbers.

- Edit the file `~/.bashrc` to include the line `source <path>/bashrc_for_gmt`. If you set up `bashrc_for_gmt` in your
  home directory, this line should be `source ~/bashrc_for_gmt`.

> Here are some of the shortcuts included in `bashrc_for_gmt`:

Reviewer: ```suggestion
Here are some of the shortcuts included in [`bashrc_for_gmt`](/admin/bashrc_for_gmt):
```

Guides for debugging GMT are provided in the
[developer resources section](https://docs.generic-mapping-tools.org/dev/debug.html) of the GMT documentation.

## Using build and test aliases

The file [admin/bashrc_for_gmt](/admin/bashrc_for_gmt) contains useful aliases for building and testing GMT that some
developers chose to use. New pull requests with other aliases that you find helpful are welcome. This file is version controlled, so you should copy the file to a different location in order to edit and use it. For example, use these
commands to copy it to your home directory:

```
cd <path-to-gmt>

> cp admin/bashrc_for_gmt ~/bashrc_for_gmt

Reviewer: Not sure about bashrc best practices, but would it be better to make it a hidden file (with a dot `.`)?

```suggestion
cp admin/bashrc_for_gmt ~/.bashrc_for_gmt
```

- If you do not have [ninja](https://ninja-build.org/) installed, you will need to change `builder=ninja` to
  `builder=make` and `Bname="Ninja"` to `Bname="Unix Makefiles"`. It is recommended that you use ninja.
- You may need to update `pngview=open` and `pngview=open` depending on your preferred program for viewing files.
- Optionally, change `ncores=4` to the number of cores to use for building and running tests.
- Change `MATLAB=/Applications/MATLAB_R2019a.app` to the path for your version of the MATLAB app.
- Set `REPO_DIR` to the path that contains the GMT repository.
- Set `DATA_DIR` to the path that contains the folders `dcw-gmt-1.1.4/` and `gshhg-gmt-2.3.7/` for the dcw and gshhg
  dataset respectively. If these folders are not located in the same path, you can instead delete L25
  (`DATA_DIR=<path to directory containing GSHHG and DCW>`) and set the individual paths to the GSHHG and DCW source
  in L30 (`export GMT_GSHHG_SOURCE=${DATA_DIR}/gshhg-gmt-2.3.7`) and
  L31(`export GMT_DCW_SOURCE=${DATA_DIR}/dcw-gmt-1.1.4`).
- Edit the file `~/.bashrc` to include the line `source <path>/bashrc_for_gmt`. If you set up `bashrc_for_gmt` in your

>   home directory, this line should be `source ~/bashrc_for_gmt`.

Reviewer: Wondering if this bullet point should probably be on the top of the list. Or if it's better to have people configure things properly first.

Author: I think setting up first is better.
